scripts/build.py

```diff
@@ -552,22 +649,21 @@ def _execute_tasks(tasks, batch_size=24):
 
     Runs no more than 'batch_size' tasks at a time.
     """
-    remaining_tasks = list(tasks)
+    remaining_tasks = collections.deque(tasks)
     currently_running_tasks = []
 
     while remaining_tasks or currently_running_tasks:
```

Comment:
Instead of having the multithreading like this, I was wondering if it was possible if we change it so it looked more like what is suggested in this Stackoverflow [post](https://stackoverflow.com/a/28463266) using `ThreadPool` with `map` since it looks cleaner and is probably implemented more efficiently than what we can do at the Python level. Of course, we should only go with that only if performance is actually better.

Reply:
I was looking into multithreading vs multiprocessing earlier. I started researching that topic since I noticed `pre_commit_linter.py` used `multiprocessing` vs `backend_tests.py`'s `threading`.
I even bookmarked this page for a later read:
https://eli.thegreenplace.net/2012/01/16/python-parallelizing-cpu-bound-tasks-with-multiprocessing/
So my follow-up question here is: Here are some ways, that I can think of, to further speed up the current build process beside replacing multithreading with multiprocessing. 
1) Optimize building `/extension` just like I did it with `/templates` (I saved 5 mins per Travis task by putting in if-else branching)
2) Apply build process to `manifest.json` and copy the hashed `manifest.json` to `/build` to detect changes `/third_party/.` . This will then enable us to prevent unnecessary `/third_party/` rebuild.
Do you think we are better off pursuing these venues rather applying another parallelization technique?

Reply:
If those yield a significant improvement in speed, I'm all for it. Those might lead to more significant gains (which we can hold off to another PR) since it leads to less work at the end of the day. I was never really happy with the threading implementation in `backend_tests` and here because there was more overhead than I'm comfortable with. The case with `pre_commit_linter` was that each linting operation might have parallelization underneath it, carried a bit of baggage for each linting type, and scoped to the 3 linting types, which is why `multiprocessing` was used. The reason I brought it up was to discuss whether it was worth testing it out.

Reply:
So Pooling + Mapping is somewhat different from firing multiple threads to deal with a queue of tasks. Going to require some restructuring wherever `build_tasks`, `copy_tasks` and `execute_tasks` is called/used. I will give this a try to test things out. Thanks.

Reply:
So I have managed to get a multiprocessing version of build_task going. So far, on my local, there has yet to be a tangible performance boost with multiprocessing. I will get a Travis build ready for future prosperity.

Reply:
Hmm, okay, thanks for trying it out.

Reply:
The Travis build is not quite done but I am expecting everything to pass smoothly.
Here is the build report: https://travis-ci.org/hoangviet1993/oppia/builds/416441515
Here is the branch it is built from, so that you can see my `Pool.map()` implementation:
https://github.com/hoangviet1993/oppia/blob/40bf66ab861bb251c4516385b0bb9c6f7c588cc6/scripts/build.py#L651
Note, there is some key difference in the main function so preferably you wanna pull the whole tree down for a closer study.

Reply:
I believe things aren't faster primarily because of Travis CI hardware spec though. I do not know if the time is correct on a higher CPU count.

Reply:
I agree, I'm happy with it like this. Thanks for trying it out!